---
title: "Pancreatic Cancer Model Building"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

SIAMCAT needs a feature matrix (matrix or data.frame) 
features (in rows) samples (in columns)
metadata in a data.frame, samples as row names

```{r setup, include=FALSE}
# load library
library("tidyverse")
library("matrixStats")
library("SIAMCAT")
library("ggrepel")
library("dplyr")
library("plyr")
library("readxl")
# check siamcat version
packageVersion("SIAMCAT")

PARAM <- list()
PARAM$folder.R <- paste0(getwd(), "/")
PARAM$folder <- gsub("Rmd/", "", PARAM$folder.R)
PARAM$folder.metadata <- paste0(PARAM$folder, "metadata/")
PARAM$folder.Rdata <- paste0(PARAM$folder, "Rdata/")
PARAM$folder.data <- paste0(PARAM$folder, "data/")
PARAM$folder.results <- paste0(PARAM$folder, "results/")

# load data metag motu and genes
load(paste0(PARAM$folder.Rdata, "metag.pc.Rdata"))
load(paste0(PARAM$folder.Rdata, "metag.pc.genes.Rdata"))
# load 16s data
load(paste0(PARAM$folder.Rdata, "16S.data.otu.nonfiltered.RData"))
load(paste0(PARAM$folder.Rdata, "16S.data.asv.nonfiltered.RData"))
```
###############################################################################
Function to run siamcat with centered log transformation (clr) normalization
10 fold cross validation and 10 fold re-sampling
Confounder check based on fisher exact test
###############################################################################
```{r}
runsiamcat <- function(featTable, metaTable, fileName, case, ml, norm){
  dim(featTable)
  # create SIAMCAT object and classify
  siamcat <- siamcat(feat=featTable, meta=metaTable, label="subject_disease_status", case=case)
  
  # filter based on abundance 
  siamcat <- filter.features(siamcat, filter.method = 'abundance', 
                            cutoff=0.001, verbose=3)
  check.confounders(siamcat, fn.plot = paste0(PARAM$folder.results, 
                            fileName, 'confounders.pdf'),
                            meta.in=metatest, verbose = 3)
        
  # normalize with log.clr
  siamcat <- normalize.features(siamcat, norm.method = "log.clr", feature.type = 'filtered',
                                norm.param = list(log.n0=1e-05, sd.min.q=1))
                                      
  # compute associations 
  siamcat <- check.associations(siamcat, feature.type = 'normalized', 
                                detect.lim = 10^-5, plot.type = "quantile.box",
                                fn.plot = paste0(PARAM$folder.results, 
                                          Sys.Date(), '.', fileName,'assoc.plot.pdf'))
        
  # train model
  siamcat <- create.data.split(siamcat, num.folds =10, num.resample = 10)  
  siamcat <- train.model(siamcat, method = "lasso_ll", verbose = 2)
  siamcat <- make.predictions(siamcat)
  siamcat <- evaluate.predictions(siamcat)    
  print(siamcat@eval_data$auroc)
  # evaluation plot
  model.evaluation.plot(siamcat, fn.plot = paste0(PARAM$folder.results, Sys.Date(), '.',
                                                  fileName, 'eval.plot.pdf'))
  # interpretation plot
  model.interpretation.plot(siamcat, fn.plot = paste0(PARAM$folder.results, 
                           Sys.Date(), '.', fileName,'interpret.plot.pdf'),
                            consens.thres = 0.5,
                            detect.lim = 1e-05,
                            heatmap.type = 'zscore')
        
  # save siamcat object
  save(siamcat, file = paste0(PARAM$folder.results, fileName, 'siamcat.Rdata'))
  return(siamcat) 
}
###############################################################################

runsiamcat.pos <- function(featTable, metaTable, fileName, case){
  
  dim(featTable)
  
  # create SIAMCAT object and classify
  siamcat <- siamcat(feat=featTable, meta=metaTable, label="status", case=case)
  
  # filter based on abundance 
  siamcat <- filter.features(siamcat, filter.method = 'abundance', 
                            cutoff=0.001, verbose=3)
        
  check.confounders(siamcat, fn.plot = paste0(PARAM$folder.results, 
                            fileName, 'confounders.pdf'),
                            meta.in=metatest, verbose = 3)
        
  # normalize with log.clr
  siamcat <- normalize.features(siamcat, norm.method = log.clr, feature.type = 'filtered',
                                norm.param = list(log.n0=1e-05, sd.min.q=1))
                                      
  # compute associations 
  siamcat <- check.associations(siamcat, feature.type = 'normalized', 
                                detect.lim = 10^-5, plot.type = "quantile.box",
                                fn.plot = paste0(PARAM$folder.results, 
                                          Sys.Date(), '.', fileName,'assoc.plot.pdf'))
        
  # train model
  siamcat <- create.data.split(siamcat, num.folds =10, num.resample = 10)  

  siamcat <- train.model(siamcat, method='lasso_ll', verbose = 2,
                         perform.fs = TRUE,
                         modsel.crit = list('pr'),
                         param.fs = list(thres.fs = 50,
                                         method.fs = gFC,
                                         direction='positive'))

  siamcat <- make.predictions(siamcat)
  siamcat <- evaluate.predictions(siamcat)    
  print(siamcat@eval_data$auroc)
  # evaluation plot
  model.evaluation.plot(siamcat, fn.plot = paste0(PARAM$folder.results, Sys.Date(), '.',
                                                  fileName, 'eval.plot.pdf'))
  # interpretation plot
  model.interpretation.plot(siamcat, fn.plot = paste0(PARAM$folder.results, 
                           Sys.Date(), '.', fileName,'interpret.plot.pdf'),
                            consens.thres = 0.5, 
                            detect.lim = 1e-05,
                            heatmap.type = 'zscore')
        
  # save siamcat object
  save(siamcat, file = paste0(PARAM$folder.results, fileName, 'siamcat.Rdata'))
  return(siamcat) 
}
```

###############################################################################
Spanish MetaG Taxonomic Modelling
###############################################################################

```{r}
# choose the meta variables to test
metatest <- c("jaundice", "diabetes","asthma", "acid", "rheumatoid arthritis", 
              "library_size", "center", "periodontitis", "age", "smoking", "acid_med",
              "bilirubin_direct","directB_comp", "totalB_comp", "alcohol_status", 
              "antibiotic", "metformin","obesitiy", "salicylic_med", "allhburn",
              "probiotic","heartburn", "cholesterol", "receding gums", "coded_ca19")

# run siamcat
runsiamcat(feat.st.rel.cancer.3mg, meta.st.cancer, "pc.st.naive", "PC")
runsiamcat(feat.or.rel.cancer.3mg, meta.st.cancer, "pc.or.naive", "PC")
runsiamcat(feat.st.rel.pc.3mg, meta.st.pc, "pc.st.CPvsCTR", "Pancreatitis")
runsiamcat(feat.st.rel.pc.3mg, meta.or.pc, "pc.or.CPvsCTR", "Pancreatitis")

temp=meta.st[meta.st$status!="CTR",]
feat.st.relpdac.cp.3mg=feat.st.rel.total.3mg[,rownames(temp)]
runsiamcat(feat.st.relpdac.cp.3mg, temp, "pc.st.PDACvsCP", "PC")

# run positive mode siamcat
runsiamcat.pos(feat.st.rel.cancer.3mg, meta.st.cancer, "pc.st.constrain.enricht", "PC")
runsiamcat.pos(feat.or.rel.cancer.3mg, meta.st.cancer, "pc.or.constrain.enricht", "PC")
runsiamcat.pos(feat.st.rel.pc.3mg, meta.st.pc, "pc.st.CPvsCTR", "Pancreatitis")
runsiamcat.pos(feat.st.rel.pc.3mg, meta.or.pc, "pc.or.CPvsCTR", "Pancreatitis")
```

###############################################################################
Add confounders to metaG models
###############################################################################

```{r}
# load models
load(paste0(PARAM$folder.Rdata, "st.naive.Rdata"))
siamcat.pc <- siamcat

load(paste0(PARAM$folder.Rdata, "st.constrain.enriched.Rdata"))
siamcat.pos <- siamcat

# add confounders to model
add.meta <- function(x, n){
  x <- add.meta.pred(x, pred.names = n, verbose=3)
  x <- train.model(x, method='lasso_ll', verbose=2, perform.fs = TRUE,
                   param.fs = list(thres.fs=50, method.fs='gFC', 
                                   direction='positive'))
  x <- make.predictions(x)
  x <- evaluate.predictions(x)
  return(x)
}
# combine with naive model
siamcat.jau <- add.meta(siamcat.pc, 'jaundice')
siamcat.col <- add.meta(siamcat.pc, 'cholesterol')
siamcat.dia <- add.meta(siamcat.pc, 'diabetes')
siamcat.bil <- add.meta(siamcat.pc, 'bilirubin_direct')
siamcat.met <- add.meta(siamcat.pc, 'metformin')
siamcat.smo <- add.meta(siamcat.pc, 'smoking')
siamcat.alc <- add.meta(siamcat.pc, 'alcohol_status')

# combine with constrained-enriched model 
siamcat.jau.pos <- add.meta(siamcat.pos, 'jaundice')
siamcat.col.pos <- add.meta(siamcat.pos, 'cholesterol')
siamcat.dia.pos <- add.meta(siamcat.pos, 'diabetes')
siamcat.bil.pos <- add.meta(siamcat.pos, 'bilirubin_direct')
siamcat.met.pos <- add.meta(siamcat.pos, 'metformin')
siamcat.smo.pos <- add.meta(siamcat.pos, 'smoking')
siamcat.alc.pos <- add.meta(siamcat.pos, 'alcohol_status')

model.evaluation.plot('constrained-enrichment model'= siamcat.pos,
                      'naive model'= siamcat.pc,
                      'jaundice'= siamcat.jau,
                      'billirubin'= siamcat.bil,
                      "diabetes"= siamcat.dia,
                      'cholesterol'= siamcat.col,
                      'metformin'= siamcat.met,
                      'smoking'= siamcat.smo,
                      'alcohol'= siamcat.alc,
                      'metformin'= siamcat.met,
                      fn.plot = paste0(PARAM$folder.results,  Sys.Date(),
                                       'confounders.interpret.pdf'))

# save confounder Rdata
save(siamcat.pc, siamcat.bil, siamcat.col, siamcat.dia, siamcat.jau, siamcat.alc, 
     siamcat.met, siamcat.c19, siamcat.smo, siamcat.dia.pos, siamcat.jau.pos, siamcat.pos,
     file = paste0(PARAM$folder.files, Sys.Date(), 'es.confounder.RData'))
```

###############################################################################
Add CA19-9 data to microbiome data
###############################################################################

```{r}
#reassign ca19 
a=meta(siamcat.pc)
meta.st.cancer$CA19.new <- as.numeric(gsub("<", "", meta.st.cancer$CA19))
meta(siamcat.pc) <- a
meta(siamcat.pos) <- meta(siamcat.pc)
# function for AND and OR combination
.f_and_or_combination <- function(x){
  require("tidyverse")
  require("pROC")
  require(SIAMCAT)
  stopifnot(class(x) == 'siamcat')
  stopifnot('coded_ca19' %in% colnames(meta(x)))
  stopifnot(!is.null(eval_data(x, verbose=0)))
  temp <- eval_data(x)$roc

  # compute for AND combination and OR combination separately
  pred <- enframe(rowMeans(pred_matrix(x)),
                  name='Sample_ID', value="pred_status") %>% 
    full_join(tibble(Sample_ID=rownames(meta(x)),
                     ca_status=meta(x)$coded_ca), by='Sample_ID') %>% 
    full_join(enframe(label(x)$label, 
                      name='Sample_ID', value='Group'), by='Sample_ID')
  pred$ca_status <- as.numeric(pred$ca_status)
  # AND combo
  combo <- pred %>% 
    filter(ca_status!=0) %>% 
    mutate(ca_status=case_when(ca_status==-1~0, TRUE~ca_status)) %>% 
    mutate(combo=pred_status*ca_status)
  roc.and.red <- roc(predictor=combo$combo, response=combo$Group, 
                     levels = c(-1, 1), direction = '<')
  
  combo <- pred %>% 
    mutate(ca_status=case_when(ca_status==-1~0, TRUE~ca_status)) %>% 
    mutate(combo=pred_status*ca_status)
  roc.and <- roc(predictor=combo$combo, response=combo$Group, 
                 levels = c(-1, 1), direction = '<')
  
  # OR combo
  combo <- pred %>% 
    filter(ca_status!=0) %>% 
    mutate(ca_status=case_when(ca_status==-1~0, TRUE~ca_status)) %>% 
    mutate(combo=pred_status+ca_status)
  roc.or.red <- roc(predictor=combo$combo, response=combo$Group, 
                    levels = c(-1, 1), direction = '<')
  
  combo <- pred %>% 
    mutate(ca_status=case_when(ca_status==-1~0, TRUE~ca_status)) %>% 
    mutate(combo=pred_status+ca_status)
  roc.or <- roc(predictor=combo$combo, response=combo$Group, 
                levels = c(-1, 1), direction = '<')
  
  df.res <- list('or'=roc.or, 'or.red'=roc.or.red, 
                 'and'=roc.and, 'and.red'=roc.and.red)
  return(df.res)
}
# combine Ca19-9 with microbiome data
sc.ca19=.f_and_or_combination(siamcat.pc)
sc.ca19.pos=.f_and_or_combination(siamcat.pos)
save(sc.ca19.pos,sc.ca19, file = paste0(PARAM$folder.results, 
                                        'ca19-9.combined.Rdata'))

# plot ROCs
get.roc <- function(sc, name){
  roc.obj <- sc$or
  df.return <- tibble(sensitivity=roc.obj$sensitivities,
                      specificitiy=roc.obj$specificities,
                      name=name,
                      auroc=roc.obj$auc)
}

df.plot <- bind_rows(get.roc(sc.ca19, 'OR'),
                     get.roc(sc.ca19.pos, 'OR.pos'))

p <- df.plot %>% 
  arrange(sensitivity) %>% 
  mutate(label=paste0(name, ' ', sprintf(fmt='%.2f', auroc))) %>% 
  ggplot(aes(x=specificitiy, y=sensitivity, col=label)) + 
  geom_abline(intercept = 1, slope = 1, col='black', lty=3) +
  geom_line() + 
  #scale_color_tableau(name='') + 
  xlab('Specificity') + 
  ylab('Sensitivity') + 
  theme_bw() + 
  scale_x_reverse() +
  theme(panel.grid.minor = element_blank(),
        aspect.ratio = 1,
        legend.position = c(0.65, 0.2), legend.title = element_blank())
# save the plots
ggsave(p, filename = paste0(PARAM$folder.results, 'roc.ca19.pdf'), width = 5.5, height = 4.5)
```

###############################################################################
Calculate TPR of all models
###############################################################################

```{r}
# TRP rate for different stages
calculate.tpr <- function(x){
  siamcat=x
  pred <- rowMeans(pred_matrix(siamcat))
  threshold <- siamcat@eval_data$roc$thresholds[
    which(siamcat@eval_data$roc$specificities > 0.90)[1]]
  
  pred=as.data.frame(pred)
  rownames(pred) <- colnames(siamcat@phyloseq@otu_table)
  pred = merge(pred, meta.st.cancer, by="row.names", all.x=TRUE)
  table(pred$stage)
  
  # combine stage 1+2 and 3+4
  pred <- pred %>%  
    mutate(stage.n=dplyr::recode(stage,  "1"=1,"2"=1,"3"=3,"4"=3))
  
  stage=pred$stage.n
  df.tpr <- tibble(pred=pred$pred, Group=(label(siamcat))[[1]], stage=stage)
  df.tpr <- df.tpr %>% 
    group_by(stage) %>% 
    dplyr::summarise(n=n(), 
                     pred.pos=sum(pred>threshold), 
                     pred.neg=sum(pred < threshold)) %>% 
    ungroup() %>% 
    mutate(tpr=pred.pos/n)
}
# naive mdoels
df.tpr.m1 = calculate.tpr(siamcat.pc)
df.tpr.m1$model = "Model 1"

df.tpr.m2 = calculate.tpr(siamcat.pos)
df.tpr.m2$model = "Model 2"

# repeat for combinations
calculate.tpr <- function(x){
  siamcat=x
  pred <- siamcat$or$predictor
  threshold <- siamcat$or$thresholds[which(siamcat$or$specificities > 0.90)[1]]
  
  pred=as.data.frame(pred)
  rownames(pred) <- colnames(siamcat.pc@phyloseq@otu_table)
  pred = merge(pred, meta.st.cancer, by="row.names", all.x=TRUE)
  table(pred$stage)
  
  pred <- pred %>%  
    mutate(stage.n=dplyr::recode(stage,  "1"=1,"2"=1,"3"=3,"4"=3))
  
  stage=pred$stage.n
  
  df.tpr <- tibble(pred=pred$pred, Group=(label(siamcat.pc))[[1]], stage=stage)
  df.tpr <- df.tpr %>% 
    group_by(stage) %>% 
    dplyr::summarise(n=n(), 
                     pred.pos=sum(pred>threshold), 
                     pred.neg=sum(pred < threshold)) %>% 
    ungroup() %>% 
    mutate(tpr=pred.pos/n)
}

# CA19-9 combined models
df.tpr.ca19.m1 = calculate.tpr(sc.ca19)
df.tpr.ca19.m1$model = "Model 1 + CA19-9"

df.tpr.c19.m2 = calculate.tpr(sc.ca19.pos)
df.tpr.c19.m2$model = "Model 2 + CA19-9"

#combine all tpr rate for plotting
df= rbind(df.tpr.m1, df.tpr.m2, df.tpr.ca19.m1, df.tpr.c19.m2)

# add CP cases from model
df[13:14,] <- ""
df$stage[13:14] <- c("CP", "CP")
df$tpr[13:14] <- c(0.14, 0.17)
df$model[13:14] <- c("Model 1","Model 2")
df$tpr<-as.numeric(df$tpr)

# plot tpr
p.tpr <- ggplot(data=df, aes(x=stage, y=tpr, fill=stage)) + 
  geom_col(alpha=0.8) + 
  theme_classic()  +  facet_grid(cols = vars(model), scales = "free") 

ggsave(p.tpr, filename=paste0(PARAM$folder.results, "stage.tpr.barplot.pdf"), 
       width = 5, height=6)
```

###############################################################################
Spanish 16S rRNA Modelling
###############################################################################

```{r}
# 16S modelling
meta.16 = amplicon %>%
  filter(str_detect(sample_alias, '^MMPC-310')) %>%
  filter(subject_disease_status!='Pancreatitis') 
# split sites
meta.16.st =  meta.16 %>%
  filter(str_detect(experiment_name, 'ST$') )%>% 
  remove_rownames %>% 
  column_to_rownames(var="experiment_name")
meta.16.or =  meta.16 %>%
  filter(str_detect(experiment_name, 'OR$')) %>%
  remove_rownames %>% 
  column_to_rownames(var="experiment_name")

# open-ref
data.otu.rel= prop.table(as.matrix(data.otu), 2)
st.16s = data.otu.rel[, match(rownames(meta.16.st), colnames(data.otu.rel))]
or.16s= data.otu.rel[, match(rownames(meta.16.or), colnames(data.otu.rel))]
dim(st.16s)
dim(or.16s)
runsiamcat(st.16s, meta.16.st, "st.16s.", "PC")
runsiamcat(or.16s, meta.16.or, "or.16s.", "PC")

# asv
data.asv.rel=prop.table(as.matrix(data.asv), 2)
st.16s = data.asv.rel[, match(rownames(meta.16.st),colnames(data.asv.rel))]
or.16s = data.asv.rel[, match(rownames(meta.16.or),colnames(data.asv.rel))]
dim(st.16s)
dim(or.16s)
runsiamcat(st.16s, meta.16.st, "st.16s.asv.", "PC")
runsiamcat(or.16s, meta.16.or, "or.16s.asv.", "PC")
```

```{r session_info}
sessionInfo()
```


